### PR TITLE
Added support for Data Repository Associations with Pcluster Managed FSx for Lustre

### PR DESCRIPTION
**Pull request by hgreebe**
### Description of changes
* FSx released Persistent_2 which no longer supports the use of ExportPath and ImportPath. Instead, it uses DRAs to configure S3 association. 
* The possibility to configure DRAs for managed FSx for Lustre file systems when configuring a cluster will now be supported.

### Tests
- Integration Tests
   -The addition, removal, and update of DRAs was tested
* Unit Tests
  -The new schema and validators were tested

### References
* https://github.com/aws/aws-parallelcluster/issues/4499

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -5,6 +5,7 @@ CHANGELOG
 ------
 
 **ENHANCEMENTS**
+- Add support for Data Repository Associations with Pcluster Managed FSx for Lustre.
```
Comment (enrico-usai) [inline CHANGELOG.md]: - Add support for Data Repository Associations when using `PERSISTENT_2` as `DeploymentType` for a managed FSx for Lustre.

File: cli/src/pcluster/config/config_patch.py [hunk 2]
```diff
@@ -189,18 +189,19 @@ def _compare_list(self, base_section, target_section, param_path, data_key, fiel
...
-            if not base_nested_section.get("visited", False):
-                self.changes.append(
-                    Change(
-                        param_path,
-                        data_key,
-                        base_nested_section,
-                        None,
-                        change_update_policy,
-                        is_list=True,
+        if base_section:
```
Comment (hanwen-pcluste) [inline cli/src/pcluster/config/config_patch.py:192]: Why this change?

File: tests/integration-tests/tests/storage/test_fsx_lustre.py [hunk 3]
```diff
@@ -179,6 +180,49 @@ def test_fsx_lustre(
     )
 
 
+@pytest.mark.usefixtures("os", "instance", "scheduler")
+def test_fsx_lustre_dra(
```
Comment (hanwen-pcluste) [inline tests/integration-tests/tests/storage/test_fsx_lustre.py:184]: Have you thought about merging this test to another test?
If merging the test is hard, we can keep this as separate. If we keep this as separate, let's add it to https://github.com/aws/aws-parallelcluster/blob/develop/tests/integration-tests/configs/common/common.yaml so that the test will be included in the daily integration test

File: cli/src/pcluster/config/cluster_config.py [hunk 4]
```diff
@@ -431,6 +432,38 @@ def existing_dns_name(self):
...
 
+class DataRepositoryAssociation(Resource):
+    """Represent the Data Repository Association resource."""
+
+    def __init__(
+        self,
+        name: str = None,
+        batch_import_meta_data_on_create: bool = None,
+        data_repository_path: str = None,
+        file_system_path: str = None,
```
Comment (hanwen-pcluste) [inline cli/src/pcluster/config/cluster_config.py]: Let's not put `None` defaults for "required" parameters. This is not a bug. But it is the custom in this file.

File: cli/src/pcluster/validators/fsx_validators.py [hunk 5]
```diff
@@ -48,6 +48,28 @@ def _validate(
...
+    Verify compatibility of given S3 association options for FSX.
+    """
+
+    def _validate(self, data_repository_associations, import_path, export_path):
+        if data_repository_associations and (import_path or export_path):
+            self._add_failure(
+                "When specifying data repository associations, import and export path "
+                "can not be used on the same file system.",
+                FailureLevel.ERROR,
+            )
```
Comment (hanwen-pcluste) [inline cli/src/pcluster/validators/fsx_validators.py:64]: Is this validation already done in `validate_s3_association_unsupported_parameters`?

File: tests/integration-tests/tests/storage/test_fsx_lustre.py [hunk 6]
```diff
@@ -179,6 +180,49 @@ def test_fsx_lustre(
...
+    region,
+    pcluster_config_reader,
+    s3_bucket_factory,
+    clusters_factory,
+    test_datadir,
+    scheduler_commands_factory,
+):
+    """Test Data Respository Associations for a FSx Lustre file system."""
+    mount_dir = "/fsx_mount_dir"
+    bucket_name = "integ-tests-knffdu7meov017vf"  # s3_bucket_factory()
```
Comment (hanwen-pcluste) [inline tests/integration-tests/tests/storage/test_fsx_lustre.py]: Use `s3_bucket_factory()`

File: tests/integration-tests/tests/storage/test_fsx_lustre/test_fsx_lustre_dra/pcluster.config.yaml [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    KeyName: {{ key_name }}
+  Iam:
+    S3Access:
+      - BucketName: {{ bucket_name }}
+        EnableWriteAccess: False
+  Imds:
+    Secured: {{ imds_secured }}
+Scheduling:
+  Scheduler: {{ scheduler }}
+  {% if scheduler == "awsbatch" %}AwsBatchQueues:{% else %}SlurmQueues:{% endif %}
```
Comment (hanwen-pcluste) [inline tests/integration-tests/tests/storage/test_fsx_lustre/test_fsx_lustre_dra/pcluster.config.yaml]: We don't need this conditional statement, because this this is for Slurm only anyway. Same comment for pcluster.config.update.yaml

File: tests/integration-tests/tests/storage/test_fsx_lustre/test_fsx_lustre_dra/pcluster.config.yaml [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    - Name: queue-0
+      Iam:
+        S3Access:
+          - BucketName: {{ bucket_name }}
+            EnableWriteAccess: False
+      ComputeResources:
+        - Name: compute-resource-0
+          {% if scheduler == "awsbatch" %}
```
Comment (hanwen-pcluste) [inline tests/integration-tests/tests/storage/test_fsx_lustre/test_fsx_lustre_dra/pcluster.config.yaml]: We don't need this conditional statement, because this this is for Slurm only anyway. Same comment for pcluster.config.update.yaml

File: tests/integration-tests/tests/storage/test_fsx_lustre.py (hunk 3 above)
Comment (hgreebe) [reply tests/integration-tests/tests/storage/test_fsx_lustre.py:184]: I considered adding it to test_fsx_luster but since that uses config files with import/export path it got a bit complicated because I would have to unmount the old file system and then mount a new one with dras

File: cli/src/pcluster/config/config_patch.py (hunk 2 above)
Comment (hgreebe) [reply cli/src/pcluster/config/config_patch.py:192]: because when the _test_storage from test_config_patch.py runs, it calls _compare_list on with a base_section of None, because it tries to go into data_repository_association since it is a nested attribute, but data_repository_association is None, which makes the test fail without this change.

File: tests/integration-tests/configs/common/common.yaml [hunk 9]
```diff
@@ -540,6 +540,16 @@ storage:
...
+  test_fsx_lustre.py::test_fsx_lustre_dra:
+    dimensions:
+      - regions: [ "eu-west-2" ]
+        instances: {{ common.INSTANCES_DEFAULT_X86 }}
+        oss: [ "alinux2" ]
+        schedulers: [ "slurm" ]
+      - regions: [ "eu-north-1" ]
+        instances: {{ common.INSTANCES_DEFAULT_ARM }}
+        oss: [ "ubuntu2204" ]
+        schedulers: [ "slurm" ]
```
Comment (hanwen-pcluste) [inline tests/integration-tests/configs/common/common.yaml]: We can have the test run on a single region with a single os and instance type. Because this feature doesn't depends on os and instance type.
We can use eu-north-1

File: cli/src/pcluster/config/config_patch.py (hunk 2 above)
Comment (hanwen-pcluste) [reply cli/src/pcluster/config/config_patch.py:192]: Thank you.

**Review by hanwen-pcluste — APPROVED** [review]
